pkg/datapath/linux/node.go

```diff
@@ -1857,25 +1861,36 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
 		ipsec.DeleteXfrm()
 	}
 
+	var errs error
 	if newConfig.UseSingleClusterRoute {
-		n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv4(), newConfig.EnableIPv4)
-		n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv6(), newConfig.EnableIPv6)
+		if err := n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv4(), newConfig.EnableIPv4); err != nil {
+			errs = errors.Join(errs, err)
+		}
+		if err := n.updateOrRemoveClusterRoute(n.nodeAddressing.IPv6(), newConfig.EnableIPv6); err != nil {
+			errs = errors.Join(errs, err)
+		}
 	} else if prevConfig.UseSingleClusterRoute {
 		// single cluster route has been disabled, remove route
-		n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false)
-		n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false)
+		if err := n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false); err != nil {
+			errs = errors.Join(errs, err)
+		}
+		if err := n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false); err != nil {
+			errs = errors.Join(errs, err)
+		}
```

Comment:
I would add more context here as well:

```suggestion
		if err := n.deleteNodeRoute(n.nodeAddressing.IPv4().AllocationCIDR(), false); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to remove old IPv4 cluster route: %w", err))
		}
		if err := n.deleteNodeRoute(n.nodeAddressing.IPv6().AllocationCIDR(), false); err != nil {
			errs = errors.Join(errs, fmt.Errorf("failed to remove old IPv6 cluster route: %w", err))
		}
```

Additionally, when I think about this code path a bit more, it's a bit worrying if we couldn't delete a previous node route for the cluster when the node changed configuration. Failing to do this could override the desired routing configuration for the node and potentially cause connectivity outages. At the same time, I'm not sure how this could be possible. `UseSingleClusterRoute` comes from the Agent configuration, so if it changes, then a restart is required for it to take effect.

Let's see if @cilium/sig-datapath could explain a bit more about this path and what sort of error visibility makes sense.

Reply:
Maybe ping the original author? It's unlikely the whole datapath group (28 people) is going to know about this.

Reply:
The original author is Thomas Graf :wink:

Reply:
Then nobody knows :sweat_smile: